code/obj/artifacts/artifact_machines/heater.dm

```diff
@@ -8,40 +8,95 @@
+	deact_text = "stops emitting gas."
 	react_xray = list(10,85,80,5,"COMPLEX")
-	var/heat_target = 310
-	var/heat_amount = 40000
+	var/gas_type = "oxygen"
+	var/gas_temp = 310
+	var/gas_amount = 100
 	examine_hint = "It is covered in very conspicuous markings."
 
 	post_setup()
 		. = ..()
-		heat_target = rand(0,620)
+		// gas type
+		// oxygen is really the only thing here that's not dangerous, so I am having it be pretty common
+		switch(artitype.name)
+			if("eldritch") // bad things
+				gas_type = pick(
+					100;"nitrogen",
+					100;"plasma",
+					100;"carbon dioxide",
+					75;"sleeping agent")
+			if("martian") // organic stuff
+				gas_type = pick(
+					200;"oxygen",
+					100;"nitrogen",
+					50;"carbon dioxide",
+					50;"farts")
+			if("ancient") // industrial type stuff
+				gas_type = pick(
+					200;"oxygen",
+					100;"nitrogen",
+					50;"carbon dioxide",
+					50;"agent b")
+			if("precursor") // the rest
+				gas_type = pick(
+					125;"oxygen",
+					100;"nitrogen",
+					75;"plasma",
+					50;"carbon dioxide",
+					30;"farts",
+					30;"agent b",
+					30;"sleeping agent")
+
+		// temperature
+		gas_temp = rand(0,620)
 		if (artitype.name == "eldritch" && prob(66))
-			if (heat_target > 310)
-				heat_target *= 2
-			if (heat_target < 310 && heat_target != 0)
-				heat_target /= 2
-		heat_amount = rand(20000,60000)
-		if (heat_target > 310)
-			activ_text = "begins to radiate hot air!"
-			deact_text = "stops pumping out hot air."
-		if (heat_target < 310)
-			activ_text = "begins to radiate cold air!"
-			deact_text = "stops pumping out cold air."
+			if (gas_temp > 310)
+				gas_temp *= 2
+			if (gas_temp < 310)
+				gas_temp /= 2
+
+		// amount
+		gas_amount = rand(50,200)
+
+		// text
+		var/tempText = ""
+		if (gas_temp > 310)
+			tempText = "hot "
+		else if (gas_temp < 310)
+			tempText = "cold "
+
+		src.activ_text = "begins to emit [tempText] gas!"
+		src.deact_text = "stops emitting [tempText] gas."
 
 	effect_process(var/obj/O)
 		if (..())
 			return
 		var/turf/simulated/L = get_turf(O)
 		if(istype(L))
-			var/datum/gas_mixture/env = L.return_air()
-			if(env.temperature < (heat_target+T0C))
-				var/transfer_moles = 0.25 * TOTAL_MOLES(env)
-				var/datum/gas_mixture/removed = env.remove(transfer_moles)
-				if(removed)
-					var/heat_capacity = HEAT_CAPACITY(removed)
-					if(heat_capacity)
-						removed.temperature = (removed.temperature*heat_capacity + heat_amount)/heat_capacity
-				env.merge(removed)
+			var/datum/gas_mixture/gas = unpool(/datum/gas_mixture)
+			gas.zero()
+			switch(src.gas_type)
+				if("oxygen")
+					gas.oxygen = src.gas_amount
+				if("nitrogen")
+					gas.nitrogen = src.gas_amount
+				if("plasma")
+					gas.toxins = src.gas_amount
+				if("carbon dioxide")
+					gas.carbon_dioxide = src.gas_amount
+				if("farts")
+					gas.farts = src.gas_amount
+				if("agent b")
+					var/datum/gas/oxygen_agent_b/trace = new
+					gas.trace_gases = list(trace)
```

Comment:
This seems unnecessary to me, since I unpool the gas mixture right before that. Surely there wouldn't be any trace gas in it?

Reply:
It technically isn't required because the gas is immediately assumed which gives the will properly maintain the associative list. More a fear of some one else copying-pasta the pattern and breaking things.

Reply:
That does make sense, I mean, I stole it from molitz b^^